### ISSUE TITLE
7X gpinitsystem does not honor --encoding which affects upgrading from 6->7

### ISSUE THREAD

**Issue by kalensk** [greenplum-db/gpdb#14330]
## Bug Report

A manual initdb honors `-E UTF8`:
```
 /usr/local/gpdb7/bin/initdb -E UTF8 -D /tmp/master        --max_connections=250 --shared_buffers=128000kB --data-checksums
The files belonging to this database system will be owned by user "gpadmin".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.UTF-8".
The default text search configuration will be set to "english".
```

gpinitsystem does not seem to not honor  `-E UTF8`:

Executing: 
`/usr/bin/bash -c source /usr/local/gpdb7/greenplum_path.sh && /usr/local/gpdb7/bin/gpinitsystem -a -I /home/gpadmin/.gpupgrade/gpinitsystem_config`

With the following gpinitsystem_config:
```
ARRAY_NAME="gp_upgrade cluster"
SEG_PREFIX=demoDataDir.MNBlKjlhQyE.
TRUSTED_SHELL=ssh
HBA_HOSTNAMES=0
ENCODING=UTF8
CHECK_POINT_SEGMENTS=8
QD_PRIMARY_ARRAY=gpadmin-laptop~gpadmin-laptop~50432~/home/gpadmin/workspace/gpdb6/gpAux/gpdemo/datadirs/qddir/demoDataDir.MNBlKjlhQyE.-1~1~-1
declare -a PRIMARY_ARRAY=(
	gpadmin-laptop~gpadmin-laptop~50434~/home/gpadmin/workspace/gpdb6/gpAux/gpdemo/datadirs/dbfast1/demoDataDir.MNBlKjlhQyE.0~2~0
	gpadmin-laptop~gpadmin-laptop~50435~/home/gpadmin/workspace/gpdb6/gpAux/gpdemo/datadirs/dbfast2/demoDataDir.MNBlKjlhQyE.1~3~1
	gpadmin-laptop~gpadmin-laptop~50436~/home/gpadmin/workspace/gpdb6/gpAux/gpdemo/datadirs/dbfast3/demoDataDir.MNBlKjlhQyE.2~4~2
```

Shows the following in the logs:
```
20221024:18:11:24:2986330 gpinitsystem:gpadmin-laptop:gpadmin-[INFO]:-Commencing local /usr/local/gpdb7/bin/initdb -E UTF8 -D /home/gpadmin/workspace/gpdb6/gpAux/gpdemo/datadirs/qddir/demoDataDir.MNBlKjlhQyE.-1        --max_connections=250 --shared_buffers=128000kB --
data-checksums
The files belonging to this database system will be owned by user "gpadmin".
This user must also own the server process.

The database cluster will be initialized with locale "C".
The default text search configuration will be set to "english".

....

20221024:18:11:26:2988255 gpcreateseg.sh:gpadmin-laptop:gpadmin-[INFO][2]:-Processing segment gpadmin-laptop
The files belonging to this database system will be owned by user "gpadmin".
This user must also own the server process.

The database cluster will be initialized with locale "C".
The default text search configuration will be set to "english".
```

This results in pg_upgrade failing with:
```
c_collate values for database "postgres" do not match:  old "en_US.utf8", new "C"
lc_ctype values for database "postgres" do not match:  old "en_US.utf8", new "C"
lc_collate values for database "template1" do not match:  old "en_US.utf8", new "C"
lc_ctype values for database "template1" do not match:  old "en_US.utf8", new "C"
```

### Greenplum version or build
7.0.0-alpha.0+dev.15878.g5fcb4ad9ad build dev'

**Comment by bmdoil** [comment]
The locale isn't being set in this case when using the input file option for gpinitsystem

From what I can tell, `LOCALE_SETTING` can be set explicitly in 3 ways.
1. `gpinitsystem -n en_US.UTF8`
2. `gpinitsystem --locale=`en.US.UTF8`
3.  Adding `LOCALE_SETTING` to the config file.

The input file option generates a config file that's then used to bootstrap, **but it has no option to set locale**.

In this case we see the log
```
20231208:13:35:25:554459 gpinitsystem:bdoil-ub:bdoil-[INFO]:-Locale has not been set in , will set to default value
20231208:13:35:25:554459 gpinitsystem:bdoil-ub:bdoil-[INFO]:-Dump current system locale to log file
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```
It falls back to using those values from the `locale` output.

BUT, if the `LANG` env variable is unset, `locale` falls back to `POSIX`

```
LANG=
LANGUAGE=
LC_CTYPE="POSIX"
LC_NUMERIC="POSIX"
LC_TIME="POSIX"
LC_COLLATE="POSIX"
LC_MONETARY="POSIX"
LC_MESSAGES="POSIX"
LC_PAPER="POSIX"
LC_NAME="POSIX"
LC_ADDRESS="POSIX"
LC_TELEPHONE="POSIX"
LC_MEASUREMENT="POSIX"
LC_IDENTIFICATION="POSIX"
LC_ALL=
```
initdb checks local via `setlocale`
From `man setlocale` 
```
If locale is an empty string, "", each part of the locale that should be modified is set according to the environment variables.  The details are implementation-dependent.  For glibc, first (regardless of category), the  environment  variable LC_ALL is inspected, next the environment variable with the same name as the category (see the table above), and finally the environment variable LANG.  The first existing environment variable is used.  If its value is not a valid locale specification, the locale is unchanged, and setlocale() returns NULL.

The locale "C" or "POSIX" is a portable locale; it exists on all conforming systems.
```
So this is how we get the `C` locale. 

From what I can tell `gpinitsystem` has behaved like this forever. The reason it's popping up in gpupgrade is because we clear most of the env vars to avoid environment mixing. 

This commit https://github.com/greenplum-db/gpupgrade/pull/897/commits/5347797e4e9eb4b2dcf22aa8c4904195b2b7563a resolves the issue for gpupgrade. 

This upstream commit https://github.com/postgres/postgres/commit/9637badd9f9209166140eb567602e91699dd2ffb will copy source cluster locale to target cluster, which will avoid any headaches with environment.

tldr: upgrade can work around this in a few ways, the easiest of which is preserving `LANG`. We can leave this open if @greenplum-db/cluster-management has interest in fixing the input config interaction, but since upgrade is unblocked we're not going to pursue it further.

**Comment by kalensk** [comment]
Thanks @bmdoil for the detailed an very clear breakdown of what is going on! Super easy to get things mixed up or confused amongst all that. 

Agreed that we should try and backport that upstream commit (and depending see if we can keep things clean and simple and undo the gpupgrade workaround.)

**Comment by nimish350** [comment]
@greenplum-db/cluster-management Will take this up for fixing it in gpinitsystem.
